Source/DafnyCore/ProofDependencyWarnings.cs

```diff
@@ -95,6 +143,106 @@ private static void Warn(DafnyOptions dafnyOptions, ErrorReporter reporter, Proo
         }
       }
     }
+
+    if ((options.Get(CommonOptionBag.SuggestProofRefactoring) || options.Get(CommonOptionBag.AnalyseProofs)) && manager.idsByMemberName[scopeName].Decl is Method m) {
+      SuggestFunctionHiding(unusedFunctions, m);
+      SuggestByProofRefactoring(scopeName, assertCoverage.ToList());
+    }
+  }
+
+  private static void SuggestFunctionHiding(List<Function> unusedFunctions, Method m) {
+    if (unusedFunctions.Any()) {
+      reporter.Info(MessageSource.Verifier, m.Body.StartToken,
+        $"Consider hiding {(unusedFunctions.Count > 1 ? "these functions, which are" : "this function, which is")} unused by the proof: {unusedFunctions.Comma()}");
+    }
+  }
+
+  private static void SuggestByProofRefactoring(string scopeName,
+    IReadOnlyList<DafnyConsolePrinter.VerificationRunResultPartialCopy> verificationRunResults) {
+    foreach (var (dep, lAsserts) in ComputeAssertionsUsedByFact(scopeName, verificationRunResults)) {
+      var factIsOnlyUsedByOneAssertion = lAsserts.Count == 1;
+      if (!factIsOnlyUsedByOneAssertion) {
+        continue;
+      }
+
+      DafnyConsolePrinter.AssertCmdPartialCopy cmd = null;
+      foreach (var assert in lAsserts) {
+        cmd = assert;
+      }
+
+      manager.ProofDependenciesById.TryGetValue(cmd.Id, out var consumer);
+
+      if (consumer != null && (dep == consumer || consumer.Range.Intersects(dep.Range))) {
```

Comment:
Could you explain what `consumer.Range.Intersects(dep.Range)` does?

Reply:
If the two overlap one is already in the by-proof of the other, so there's nothing to be gained.